File: WebCatalogBackend/src/main/java/com/cafebabe/model/mobilephone/MobilePhoneModel.java
package com.cafebabe.model.mobilephone;

import com.cafebabe.model.*;
import lombok.Getter;
import lombok.Setter;

import javax.persistence.*;
import java.util.List;

@Getter
@Setter
@Entity
public class MobilePhoneModel extends BaseDataObject {

    @ManyToOne
    protected Manufacturer manufacturer;

    protected int marketLaunchYear;

    @Enumerated(EnumType.STRING)
    protected MobilePhoneType mobilePhoneType;

    @ManyToOne
    protected OperationSystemWithVersion operationSystem;

    protected Float screenDiagonalInInches;

    @OneToMany
    protected List<RamAndRomVariant> ramAndRomVariants;

    protected Integer horizontalScreenResolution;

    protected Integer verticalScreenResolution;

    @ManyToOne
    protected ScreenTechnology screenTechnology;

    protected Integer screenRefreshRate;

    @ManyToOne
    protected RamType ramType;

    @ManyToOne
    protected FingerprintScannerLocation fingerprintScannerLocation;

    @ManyToOne
    protected RomType romType;

    protected Integer simCardsAmount;

    @ManyToOne
    protected SimCardType simCardType;

    protected boolean isMemoryCardSupported;

    protected Integer camerasAmount;

    protected Integer cameraInMp;

    protected boolean is5GSupported;

//    @ManyToMany
//    protected List<MemoryCardType> supportedMemoryCardTypes;

//    protected Integer maxMemoryCardSizeInGb;

    protected boolean hasAudioProcessor;

    protected float frontCamera;

    protected boolean hasAudioOutput;

    @ManyToOne
    protected ConnectionSocket connectionSocket;

    protected String bodyMaterial;

    protected Float length;

    protected Float width;

    protected Float thickness;

    protected Float weight;
}

File: WebCatalogFrontend/src/app/model/mobile-phone-model.ts
import {BaseDataObject} from "./base-data-object";
import {Manufacturer} from "./manufacturer";
import {OperationSystemWithVersion} from "./operation-system-with-version";
import {ScreenTechnology} from "./screen-technology";
import {RamAndRomVariant} from "./ram-and-rom-variant";
import {DustAndMoistureProtection} from "./dust-and-moisture-protection";
import {FingerprintScannerLocation} from "./fingerprint-scanner-location";
import {Cpu} from "./cpu";

export class MobilePhoneModel extends BaseDataObject {

  manufacturer: Manufacturer;
  marketLaunchYear: number;
  operationSystem: OperationSystemWithVersion;
  screenDiagonalInInches: number;
  horizontalScreenResolution: number;
  verticalScreenResolution: number;
  screenTechnology: ScreenTechnology;
  screenRefreshRate: number;
  ramAndRomVariants: Array<RamAndRomVariant>;
  isMemoryCardSupported: boolean;
  camerasAmount: number;
  cameraInMp: number;
  simCardsAmount: number;
  simCardType: BaseDataObject;
  is5GSupported: boolean;
  bodyColor: string;
  dustAndMoistureProtection: DustAndMoistureProtection;
  batteryCapacity: number;
  fingerprintScannerLocation: FingerprintScannerLocation;
  screenResolutionPpi: number;
  screenProtection: BaseDataObject;
  cpu: Cpu;
  hasAudioProcessor: boolean;
  frontCameraInMp: number;
  hasAudioOutput: boolean;
  connectionSocket: BaseDataObject;
  length: number;
  width: number;
  thickness: number;
  weight: number;

  constructor(name: string, manufacturer: Manufacturer, marketLaunchYear: number, operationSystem: OperationSystemWithVersion, screenDiagonalInInches: number, horizontalScreenResolution: number, verticalScreenResolution: number, screenTechnology: ScreenTechnology, screenRefreshRate: number, ramAndRomVariants: Array<RamAndRomVariant>, isMemoryCardSupported: boolean, camerasAmount: number, cameraInMp: number, simCardsAmount: number, simCardType: BaseDataObject, is5GSupported: boolean, bodyColor: string, dustAndMoistureProtection: DustAndMoistureProtection, batteryCapacity: number, fingerprintScannerLocation: FingerprintScannerLocation, screenProtection: BaseDataObject, cpu: Cpu, hasAudioProcessor: boolean, frontCameraInMp: number, hasAudioOutput: boolean, connectionSocket: BaseDataObject, length: number, width: number, thickness: number, weight: number) {
    super();
    this.name = name;
    this.manufacturer = manufacturer;
    this.marketLaunchYear = marketLaunchYear;
    this.operationSystem = operationSystem;
    this.screenDiagonalInInches = screenDiagonalInInches;
    this.horizontalScreenResolution = horizontalScreenResolution;
    this.verticalScreenResolution = verticalScreenResolution;
    this.screenTechnology = screenTechnology;
    this.screenRefreshRate = screenRefreshRate;
    this.ramAndRomVariants = ramAndRomVariants;
    this.isMemoryCardSupported = isMemoryCardSupported;
    this.camerasAmount = camerasAmount;
    this.cameraInMp = cameraInMp;
    this.simCardsAmount = simCardsAmount;
    this.simCardType = simCardType;
    this.is5GSupported = is5GSupported;
    this.bodyColor = bodyColor;
    this.dustAndMoistureProtection = dustAndMoistureProtection;
    this.batteryCapacity = batteryCapacity;
    this.fingerprintScannerLocation = fingerprintScannerLocation;
    this.screenProtection = screenProtection;
    this.cpu = cpu;
    this.hasAudioProcessor = hasAudioProcessor;
    this.frontCameraInMp = frontCameraInMp;
    this.hasAudioOutput = hasAudioOutput;
    this.connectionSocket = connectionSocket;
    this.length = length;
    this.width = width;
    this.thickness = thickness;
    this.weight = weight;
  }
}

export enum MobilePhoneType {

  SMARTPHONE = 'SMARTPHONE', SIMPLE_MOBILE_PHONE = 'SIMPLE_MOBILE_PHONE'
}
